Posts/2021/04/tfl-WMIEventHandler.md

```diff
@@ -0,0 +1,363 @@
+* **Log File Consumer**: This handler writes strings of customisable text to a text file.
+* **NT Event Log Consumer**: This consumer writes event details into the Windows Application event log.
+* **SMTP Event Consumer**: You use this consumer to send an SMTP email message when an event occurs
+* **Command Line Consumer**:  This consumer runs a program with parameters, for example, run PowerShell 7 and specific a script to run.
+
+The Active Script consumer _only_ runs VBS scripts.
+Short of redeveloping the COM component, you can not use this consumer with Powershell scripts.
+The Log File Consumer is excellent for writing short highly-customised messages to a text file but can take some time and effort to implement.
+For most IT Pros, the Command Line consumer is the one to choose.
+With this consumer, you get WMI to run a PowerShell script any time an event occurs, such as a change to an AD group.
+Let's look at how you use this permanent event consumer to discover changes to the membership of the Enterprise Admins group.
+
+## Creating a permanent event handler
+
+With WMI permanent event handling, you need to create three objects within the CIM database
+
+* Event filter - the filter tells WMI which event to detect, such as a change in the change to an AD group.
+* Event consumer - this tells WMI which permanent event consumer to run and how to invoke the consumer, such as to run the Command Line consumer and run `Monitor.ps1`.
+* Event binding - this binds the filter (what event to look out for) to the consumer (what to do when the event occurs happens)
+
+To carry out these three operations, you inserting new objects into three specific WMI system classes.
+The WMI system class instances enable WMI to continue to process events after you stop your PowerShell session, log off, or restart your host.
+
+In the code below, you use the Command Line consumer to detect changes to the AD's Enterprise Admins group.
+Every time the change event occurs, you want WMI to run a specific script, namely `Monitor.ps1`.
+This script displays a list of the current members of the Enterprise Admins group to a log file and reports whether the membership now contains unauthorised users.
+If the script finds that an unauthorised user is now a group member, it writes details to a text file for you to review later.
+
+## The Solution
+
+There are several steps in this solution.
+So please, fasten your seat belts, and away we go.
+
+### Setting up
+
+In this post, you want to detect whether an unauthorised user is a member of the Enterprise Admins group.
+You must first create a file of authorised users.
+Then you create two helper functions to assist you in testing the code.
+The function to delete all aspects of the WMI event filter from your host is useful unless you plan to keep the filter running forever.
+
+```powershell
+
+# 1. Create a list of valid users for the Enterprise Admins group
```

Comment:
Remove the leading and trailing blank lines in a code block.

```suggestion
# 1. Create a list of valid users for the Enterprise Admins group
```

Reply:
will do